primers/pcr/pcr.go

```diff
@@ -0,0 +1,184 @@
+fragments, referred to as "primers" or "oligos", can be designed on a computer
+and then synthesized for amplifying a variety of different templates.
+
+This package allows users to simulate a PCR reaction or design new primers to
+amplify a given template. This package assumes perfect annealing to template at
+a target temperature, so should only be used for PCR reactions where this is a
+reasonable assumption.
+
+If you are trying to simulate amplification out of a large pool, such as an
+oligo pool, use the `Simulate` rather than `SimulateSimple` function to detect
+if there is concatemerization happening in your multiplex reaction. In most
+other cases, use `SimulateSimple`.
+
+The targetTm in all functions is specifically for Taq polymerase.
+*/
+package pcr
+
+import (
+	"errors"
+	"github.com/TimothyStiles/poly/primers"
+	"github.com/TimothyStiles/poly/transform"
+	"index/suffixarray"
+	"sort"
+	"strings"
+)
+
+// https://doi.org/10.1089/dna.1994.13.75
+var minimalPrimerLength int = 15
+
+// DesignPrimersWithOverhangs designs two primers to amplify a target sequence,
+// adding on an overhang to the forward and reverse strand. This overhang can
+// contain additional DNA needed for assembly, like Gibson assembly overhangs
+// or GoldenGate restriction enzyme sites.
+func DesignPrimersWithOverhangs(sequence, forwardOverhang, reverseOverhang string, targetTm float64) (string, string) {
+	sequence = strings.ToUpper(sequence)
+	forwardPrimer := sequence[0:minimalPrimerLength]
+	for i := 0; primers.MeltingTemp(forwardPrimer) < targetTm; i++ {
+		forwardPrimer = sequence[0 : minimalPrimerLength+i]
+	}
+	reversePrimer := transform.ReverseComplement(sequence[len(sequence)-minimalPrimerLength:])
+	for i := 0; primers.MeltingTemp(reversePrimer) < targetTm; i++ {
+		reversePrimer = transform.ReverseComplement(sequence[len(sequence)-(minimalPrimerLength+i):])
+	}
+
+	// Add overhangs to primer
+	forwardPrimer = forwardOverhang + forwardPrimer
+	reversePrimer = transform.ReverseComplement(reverseOverhang) + reversePrimer
+
+	return forwardPrimer, reversePrimer
+}
+
+// DesignPrimers designs two primers to amplify a target sequence and only that
+// target sequence (no overhangs).
+func DesignPrimers(sequence string, targetTm float64) (string, string) {
+	return DesignPrimersWithOverhangs(sequence, "", "", targetTm)
+}
+
+// SimulateSimple simulates a PCR reaction. It takes in a list of sequences and
+// a list of primers, with support for complex multiplex reactions, produces
+// a list of all possible PCR fragments from such a reaction. It does not
+// detect concatemerization, which could be useful or very detrimental to
+// your reactions. The variable `circular` is for if the target template is
+// circular, like a plasmid.
+func SimulateSimple(sequences []string, targetTm float64, circular bool, primerList []string) []string {
+	// Set all primers to uppercase.
+	for i := range primerList {
```

Comment:
This could be a function to 'normalize' sequences. Another point is, why not use structures that we already have for sequences?

Reply:
It is not meant to normalize sequences or anything, it is only meant to simulate PCR. 

> why not use structures that we already have for sequences?
Features are not necessary for PCR simulation.

Reply:
1. rename to `i` to `primer`

2. Isaac does make a good point though. It's pretty standard practice in Go to pass a large number of arguments via a struct. Something like `PCRSimulationOptions` could work here. May not need it now but I feel like our next refactor should include things like this anyways.

Reply:
Fair, but I think this function doesn't have too many options as of right now.

Reply:
Fixed the `i` stuff.